rickshaw

```diff
@@ -0,0 +1,212 @@
+    my $json_text = "";
+    while ( <JSON_FH> ) {
+        $json_text .= $_;
+    }
+    close JSON_FH;
+    my $perl_scalar = $coder->decode($json_text) || die "Could not read JSON";;
+    return $perl_scalar;
+}
+
+foreach my $e (qw(RS_USER RS_EMAIL RS_TAGS RS_DESC)) {
+    if (exists $ENV{$e}) {
+        my $var = ($e =~ s/^RS_//);
+        $run{$var} = $ENV{$e};
+    }
+}
+
+# Process the cmdline params
+while (scalar @ARGV > 0) {
+    my $p = shift @ARGV;
+    debug_log(sprintf "processing \@ARGV, param: [%s]\n", $p);
+    my $arg;
+    my $val;
+    if ( $p =~ /^\-\-(\S+)/ ) {
+        if ( $p =~ /^\-\-(\S+)=(.*)/ ) { # '--arg=val'
+            $arg = $1;
+            $val = $2;
+        } else { # '--arg val'
+            $arg = $p;
+            $val = shift @ARGV;
+        }
+        debug_log(sprintf "processing \@ARGV, arg is: [%s], val is: [%s]\n", $arg, $val);
+    } else {
+        print "[ERROR]malformed cmdline parameter: %s\n";
+        exit 1;
+    }
+    if ($arg eq "bench-config" ) {
+        debug_log(sprintf "found bench-config param\n");
+        $bench_config_file = $val;
+        $bench_config_ref = get_json_file($bench_config_file);
+        %bench_config = %{ $bench_config_ref };
+        if (exists $bench_config{'benchmark'}) {
+            printf "Preparing to run %s\n", $bench_config{'benchmark'};
+            $run{'benchmark'} = $bench_config{'benchmark'};
+        } else {
+            print "[ERROR]benchmark was not defined in %s\n", $bench_config_file;
+            exit 1;
+        }
+    } elsif ($arg eq "bench-params") {
+        $params_ref = get_json_file($val);
+        @params = @{ $params_ref };
+    } elsif ($arg eq "dir") {
+        $base_dir = $val;
+    } elsif ($arg eq "endpoint") {
+        push(@endpoints, $val);
+    } elsif ($arg =~ /^test-order$|^tool-group$|^num-samples$|^name$|^email$|^tags$|^desc|/) {
+        $run{$arg} = $val;
+    } else {
+        printf "[ERROR]parameter not valid: %s\n";
+        exit 1;
+    }
+}
+
+# Apply defaults
+foreach my $p (keys %defaults) {
+    if (! exists $run{$p}) {
+        $run{$p} = $defaults{$p};
+    }
+}
+# Confirm we have required params
+defined $bench_config_ref || die "[ERROR]You must use --bench-config=/path/to/benchmark-config.json\n";
+defined $params_ref || die "[ERROR]Your must use --bench-params=/path/to/benchmark-params.json\n";
+
+-e $base_dir || mkdir($base_dir);
+
+# If there are no endpoints, assume 1 endpoint using the 'local' extension
+if (scalar @endpoints == 0) {
+    if (exists $bench_config{'client'} and exists $bench_config{'server'}) {
+        push(@endpoints, "local:client[1],server[1]");
+    } else {
+        push(@endpoints, "local:client[1]");
+    }
+}
+
+print "\nEndpoints that will be used:\n";
+foreach my $endpoint (@endpoints) {
+    printf "%s\n", $endpoint;
+}
+$run{'endpoints'} = \@endpoints;
+
+
+printf "\nendpoint output:\n";
+# Deploy the endpoints so they are ready to run benchmark and tools
```

Comment:
So is endpoint deployment where packages would get installed, containers/vms launched, etc.?

Reply:
Yes, and then they need to be on "stand-by", issuing their first in a list of commands, which would likely be a roadblock command signaling that they are ready to run.  The rest of the commands would be tool start/stops, roadblock commands, and benchmark executions.  I think we'll need to pass either a script or a "queue of commands" when we build the endpoint filesystem-image.  Rickshaw will be responsible for building these commands for each client & server.  I thought about having each endpoint do that, but I would like to avoid that because I don't want to duplicate that work in many endpoints.  Whatever the endpoint may be (local, ssh, k8s, osp-vm, etc), the filesystem-image should be consistent, and therefore the commands it issue should be consistent regardless of the endpoint type.  The only real difference in the commands are things like IP addrs and ports.